src/NHibernate/Util/AsyncReaderWriterLock.cs

```diff
@@ -0,0 +1,229 @@
+
+		public AsyncReaderWriterLock()
+		{
+			_writerReleaser = new Releaser(this, true);
+			_readerReleaser = new Releaser(this, false);
+			_readerReleaserTask = Task.FromResult(_readerReleaser);
+		}
+
+		internal int CurrentReaders => _currentReaders;
+
+		internal int WritersWaiting => _writersWaiting;
+
+		internal int ReadersWaiting => _readersWaiting;
+
+		internal bool Writing => _currentReaders == 0 && _writeLockSemaphore.CurrentCount == 0;
+
+		internal bool AcquiredWriteLock => _writeLockSemaphore.CurrentCount == 0;
+
+		public Releaser WriteLock()
+		{
+			if (!CanEnterWriteLock(out var waitForReadLocks))
+			{
+				_writeLockSemaphore.Wait();
+				lock (_writeLockSemaphore)
+				{
+					_writersWaiting--;
+				}
+			}
+
+			if (waitForReadLocks)
+			{
+				_readLockSemaphore.Wait();
+			}
+
+			return _writerReleaser;
+		}
+
+		public async Task<Releaser> WriteLockAsync()
+		{
+			if (!CanEnterWriteLock(out var waitForReadLocks))
+			{
+				await _writeLockSemaphore.WaitAsync().ConfigureAwait(false);
+				lock (_writeLockSemaphore)
+				{
+					_writersWaiting--;
+				}
+			}
+
+			if (waitForReadLocks)
+			{
+				await _readLockSemaphore.WaitAsync().ConfigureAwait(false);
+			}
+
+			return _writerReleaser;
+		}
+
+		public Releaser ReadLock()
+		{
+			if (CanEnterReadLock())
+			{
+				return _readerReleaser;
+			}
+
+			_waitingReadLockSemaphore.Wait();
+			ReleaseWaitingReader();
+
+			return _readerReleaser;
+		}
+
+		public Task<Releaser> ReadLockAsync()
+		{
+			return CanEnterReadLock() ? _readerReleaserTask : ReadLockInternalAsync();
+
+			async Task<Releaser> ReadLockInternalAsync()
+			{
+				await _waitingReadLockSemaphore.WaitAsync().ConfigureAwait(false);
+				ReleaseWaitingReader();
+
+				return _readerReleaser;
+			}
+		}
+
+		private bool CanEnterWriteLock(out bool waitForReadLocks)
+		{
+			waitForReadLocks = false;
+			lock (_writeLockSemaphore)
+			{
+				if (_writeLockSemaphore.CurrentCount > 0 && _writeLockSemaphore.Wait(0))
+				{
+					waitForReadLocks = _currentReaders > 0;
+					return true;
+				}
+
+				_writersWaiting++;
+			}
+
+			return false;
+		}
+
+		private void ExitWriteLock()
+		{
+			lock (_writeLockSemaphore)
+			{
+				if (_writeLockSemaphore.CurrentCount == 1)
+				{
+					throw new InvalidOperationException();
+				}
+
+				// Writers have the highest priority even if they came last
+				if (_writersWaiting > 0 || _waitingReadLockSemaphore == null)
+				{
+					_writeLockSemaphore.Release();
+					return;
+				}
+
+				if (_readersWaiting > 0)
+				{
+					_currentReaders += _readersWaiting;
+					_waitingReadLockSemaphore.Release(_readersWaiting);
+				}
+
+				_writeLockSemaphore.Release();
+			}
+		}
+
+		private bool CanEnterReadLock()
+		{
+			lock (_writeLockSemaphore)
+			{
+				if (_writersWaiting == 0 && _writeLockSemaphore.CurrentCount > 0)
+				{
+					_currentReaders++;
+
+					return true;
+				}
+
+				if (_waitingReadLockSemaphore == null)
+				{
```

Comment:
Haven't we a flaw here with current `ReleaseWaitingReader`?
What if an `ExitWriteLock` starts releasing the readers, but then a new write call causes new readers to hit this point while old waiting readers have not yet all already called `ReleaseWaitingReader`?
The new readers would then use the old semaphore, allowing some of them to run immediately. Or I am missing something?

Reply:
Nice finding! You are correct, unfortunately I wasn't able to make a test that would expose the issue, `_waitingReadLockSemaphore` was already disposed when adding new readers. In order to fix the issue I used the logic that you described in the [above comment](https://github.com/nhibernate/nhibernate-core/pull/2147/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.yml#r403746842).

Reply:
I think it is possible to reprod the issue by using thread starvation:

 - Start a writer task, lock the ReaderWriterLock with it.
 - Ensure there are more asynchronously waiting readers on the ReaderWriterLock than available threads (30 readers should be enough for that).
 - Before ending the first write blocking them, start a thread for a second write which will synchronously wait (to ensure it keeps its thread for running immediately once released) with a dedicated test semaphore before locking the ReaderWriterLock.
 - Also start some other "reader" threads which will synchronously wait with a second dedicated test semaphore before locking the ReaderWriterLock.
 - Release the first write lock. At that point, the previously readers task are all released by the semaphore, but they cannot run all due to lack of threads. Have a third test lock in their task for avoiding them to finish, and so ensuring the thread starvation remains.
 - Release the second write thread with the first test semaphore. Since it was synchronously waiting, it will still have its thread and will be able to enter the ReaderWriterLock for locking it.
 - Release the "next reader" threads with the second test semaphore. Same as for the write, they have their thread available since they were synchronously waiting. Here we have a concurrency risk for the test: maybe they will be able to reach the ReaderWriterLock and enter it before the second write. Maybe a crude Thread.Sleep before releasing them would be enough to ensure this does not happen.
 - Normally the bug should happen, those "next readers" will run through the lock despite the second writer having request the lock.
 - But also at that point, likely after another crude Thread.Sleep to ensure the "next readers" had time to reach the lock, the third test lock should be released to allow the test to finish by allowing the first reader to finish.

Reply:
Thanks for the instructions, but unfortunately I still wasn't able to create a failing test for the old version. The test that was added in the last commit reproduces the issue of reusing the `_waitingReadLockSemaphore` instance for the next readers, but due to `SemaphoreSlim.Release` always releasing the first readers the issue didn't shown up. With the old version `_waitingReadLockSemaphore` is used for the first and the next readers and disposed only after the next readers are finished reading.